parlai/core/worlds.py

```diff
@@ -975,6 +975,202 @@ def shutdown(self):
+        for w in self.worlds:
+            w.reset_metrics()
+
+    def epoch_done(self):
+        return all(w.epoch_done() for w in self.worlds)
+
+    def num_examples(self):
+        return self.world.num_examples()
+
+    def num_episodes(self):
+        return self.world.num_episodes()
+
+    def _ceil(self, n, r=4):
+        """
+        Round to the nearest multiple of 4.
+
+        TensorCores only work when a tensor is a multiple of 4 in all dimensions. This
+        means all examples cost is related to their nearest multiple of 4.
+        """
+        # round up to 4, all things are equal
+        return ((n + r - 1) // r) * r
+
+    def _score(self, obs):
+        if 'text_vec' in obs:
+            # note that all examples have a cost that is based on their
+            # nearest multiple of 4. We can therefore mix-and-match
+            # anything with the same cost for increased stochasticity,
+            # while not really wasting much padding.
+            return self._ceil(len(obs['text_vec']))
+        else:
+            return None
+
+    def parley(self):
+        # first make sure that all the worlds are processed in the queue
+        assert not self.epoch_done()
+        indices = []
+        for i in range(self._BUFFER_SIZE):
+            if self._scores[i] is not None:
+                indices.append(i)
+                continue
+            if self.worlds[i].epoch_done():
+                continue
+
+            act = self.worlds[i].agents[0].act()
+            obs = self.worlds[i].agents[1].observe(act)
+            self._obs[i] = obs
+
+            if act is None:
+                # TODO: we might need to do something special here around the end of
+                # the epoch
+                raise TypeError("This is unexpected.")
+
+            self._scores[i] = self._score(obs)
+            if self._scores[i] is not None:
+                indices.append(i)
+
+        assert len(indices) != 0, "DynamicBatchWorld ran out of data!"
+
+        for i in indices:
+            assert self._scores is not None
+
+        # sort all the indices by their score, so that we can find similarly lengthed
+        # items in O(1)
+        indices = sorted(indices, key=lambda i: (self._scores[i], random.random()))
+        # indices = sorted(indices, key=lambda i: (random.random()))
+
+        # now let's build the batch
+        batch = []
+
+        # start with a random item. index_index is the lookup into indices, but
+        # index is the actual world.
+        width = 0
+        index_index = random.randint(0, len(indices) - 1)
+
+        # we picked a random spot, but we can get better packing if we start at the
+        # last example with the same score, since we always pick move down to smaller examples.
+        while index_index < len(indices) - 1 and (
+            self._scores[indices[index_index]] == self._scores[indices[index_index + 1]]
+        ):
+            index_index += 1
```

Comment:
dumb q: why aren't we doing index_index -= 1?

Reply:
sorted for shortest to longest. shortest examples are at index 0, longest at the end. We pick a point and then count down. Here though, we find the last spot with the equivalent cost so that we can maximize packing. I actually experimented with this and in my experiments it took it from like 5.5x to 6.5x.